Fix Xcode 14.3 which supports only iOS 11.0+

kenstir opened the pull request:
### Summary

Update code to build and examples to run under Xcode 14.3.

### Detailed Changes

* Replace deprecated function AVCaptureDevice devicesWithMediaType:
  with AVCaptureDeviceDiscoverySession
* Update macOS target to 10.15 (from 10.8) to gain access to AVCaptureDeviceDiscoverySession
* Update iOS target to 11.0 (from 8.0) to compile with Xcode 14.3
* Change README to reflect the new minimum OS levels
* Change and test BarcodeScanner and BarcodeScannerSwift examples

KBEMobisys commented:
@kenstir I think the [README.md](https://github.com/zxingify/zxingify-objc#readme) should also be updated, right?
Anyway, it would be very cool if we could merge this back and release a new version in CocoaPods.

KBEMobisys commented on ZXingObjC/client/ZXCapture.m:
```diff
@@ -211,12 +211,18 @@ - (int)front {
 - (BOOL)hasFront {
-  NSArray *devices = [AVCaptureDevice devicesWithMediaType:AVMediaTypeVideo];
-  return [devices count] > 1;
+  AVCaptureDeviceDiscoverySession *captureDeviceDiscoverySession = [AVCaptureDeviceDiscoverySession discoverySessionWithDeviceTypes:@[AVCaptureDeviceTypeBuiltInWideAngleCamera]
```
This change is only available on macOS 10.15 or never. So for MacOS we would have to increase the deployment target to '10.15'. To stay as compatible as possible I would advise to drop this change again until "AVCaptureDevice" is really removed.

KBEMobisys replied:
OK, message "'devicesWithMediaType:' is deprecated: first deprecated in iOS 10.0 - Use AVCaptureDeviceDiscoverySession instead." is an error and not a warning. So, then we have two options:
- Set build setting "Treat Warnings as Errors" to No
- Set deployment target of MacOS to 10.15

kenstir replied:
My understanding is that it is deprecated in iOS 10 and *removed* in iOS 11.  And since Xcode 14.3 supports only iOS 11.0+, the "Treat Warnings as Error"=no option is not viable.

Sorry I didn't catch the macOS regression, I use only iOS builds.  Let me see if I can amend this PR to increase the deployment target on macOS.  At least then this PR will be a potential option to move forward.

kenstir replied:
I updated the deployment target to 10.15, and now it is building and unit tests passing on Xcode 14.3.1.

KBEMobisys replied:
Thanks for fixing MacOS :) 

"devicesWithMediaType:" is introduced in iOS 4.0 and first deprecated in iOS 10.0. It is still available in iOS 11 and later. It's only deprecated so far. See https://developer.apple.com/documentation/avfoundation/avcapturedevice/1390520-deviceswithmediatype

If you set "Treat Warnings as Error = no" and restart xCode it is only a warning and we only have to increase MacOs to 10.13 (which adds more compatible Macs). I tested it and also ran all the tests with success. 

My suggestion would be to keep this library as compatible as possible. But that is a decision of the library owners, I guess. :)

KBEMobisys reviewed (changes requested):
One of the changes is breaking target "OS X". See other comments

kenstir commented:
> One of the changes is breaking target "OS X". See other comments

The osx target and its unit tests are passing, now that I updated the deployment  target to 10.15.  Also I updated the README to indicate the new base level.

KBEMobisys commented on README.md:
Could you also update the Ruby example "platform :ios, '8.0'" in the Readme?

Add also the "s.ios.deployment_target" and "s.osx.deployment_target" in the podspec file?

kenstir replied:
Done!  Sorry it took so long, I got distracted.

KBEMobisys reviewed (changes requested):
Appreciate your work :)

I added some new comments in Pullrequest.

adil-hussain-84 commented:
@kenstir Are you able to make the change to the "s.ios.deployment_target" and "s.osx.deployment_target" values in the "ZXingObjC.podspec" file as requested by @KBEMobisys in the pull request [here](https://github.com/zxingify/zxingify-objc/pull/575/files/a2aa6bc1a03e629bf76c1ed32c9da7f1ed4b1c1f#r1280215032)?

kenstir reviewed (commented):
self check: pass

ale-gen commented:
Hey, these changes are exactly what I'm waiting for! When can I expect to bump deployment target to keep Xcode 14 happy?

KBEMobisys commented:
@adil-hussain-84 Could you review the pull request again? And are you able to merge it or who can do this??
Would be great to finalize this soon :)

adil-hussain-84 commented:
@KBEMobisys: Sadly I don't have write access 🙁 I'm hoping one of the individuals listed in the [Kudos](https://github.com/zxingify/zxingify-objc#kudos) section of the project's README file (@cwalcott and/or @neacao perhaps) can review the MR 🙏

KBEMobisys commented:
@benjohnde @tknoelle You are members of [zxingify](https://github.com/zxingify). Could you review and merge this PR?

We would love to see a new release with this change, so zxingify is compatible to Xcode 14.3 and above.

benjohnde commented:
> @benjohnde @tknoelle You are members of [zxingify](https://github.com/zxingify). Could you review and merge this PR?
> 
> We would love to see a new release with this change, so zxingify is compatible to Xcode 14.3 and above.

Sry for the delay, I can merge it tomorrow and also prepare a cocoapods release.

adil-hussain-84 commented:
Thank you so much @benjohnde for merging in the pull request and performing the release 🙏 We were starting to fear the worst that we'd have to fork this repository and maintain the fork.
